vendor/bat-native-ledger/src/bat/ledger/internal/promotion/promotion_util.cc

```diff
@@ -0,0 +1,284 @@
+    const ledger::PromotionType type) {
+  switch (static_cast<int>(type)) {
+    case static_cast<int>(ledger::PromotionType::UGP): {
+      return ledger::ReportType::GRANT;
+    }
+    case static_cast<int>(ledger::PromotionType::ADS): {
+      return ledger::ReportType::ADS;
+    }
+    default: {
+      NOTREACHED();
+      return ledger::ReportType::GRANT;
+    }
+  }
+}
+
+bool ParseFetchResponse(
+    const std::string& response,
+    ledger::PromotionList* list) {
+  if (!list) {
+    return false;
+  }
+
+  base::Optional<base::Value> value = base::JSONReader::Read(response);
+  if (!value || !value->is_dict()) {
+    return false;
+  }
+
+  base::DictionaryValue* dictionary = nullptr;
+  if (!value->GetAsDictionary(&dictionary)) {
+    return false;
+  }
+
+  auto* promotions = dictionary->FindKey("promotions");
+  if (promotions && promotions->is_list()) {
+    const auto promotion_size = promotions->GetList().size();
+    for (auto& item : promotions->GetList()) {
+      ledger::PromotionPtr promotion = ledger::Promotion::New();
+
+      auto* id = item.FindKey("id");
+      if (!id || !id->is_string()) {
+        continue;
+      }
+      promotion->id = id->GetString();
+
+      auto* version = item.FindKey("version");
+      if (!version || !version->is_int()) {
+        continue;
+      }
+      promotion->version = version->GetInt();
+
+      auto* type = item.FindKey("type");
+      if (!type || !type->is_string()) {
+        continue;
+      }
+      promotion->type = ConvertStringToPromotionType(type->GetString());
+
+      auto* suggestions = item.FindKey("suggestionsPerGrant");
+      if (!suggestions || !suggestions->is_int()) {
+        continue;
+      }
+      promotion->suggestions = suggestions->GetInt();
+
+      auto* approximate_value = item.FindKey("approximateValue");
+      if (!approximate_value || !approximate_value->is_string()) {
+        continue;
+      }
+      promotion->approximate_value = std::stod(approximate_value->GetString());
+
+      auto* available = item.FindKey("available");
+      if (!available || !available->is_bool()) {
+        continue;
+      }
+      if (available->GetBool()) {
+        promotion->status = ledger::PromotionStatus::ACTIVE;
+      } else {
+        promotion->status = ledger::PromotionStatus::OVER;
+      }
+
+      auto* expires_at = item.FindKey("expiresAt");
+      if (!expires_at || !expires_at->is_string()) {
+        continue;
+      }
+
+      base::Time time;
+      bool success =
+          base::Time::FromUTCString(expires_at->GetString().c_str(), &time);
+      if (success) {
+        promotion->expires_at = time.ToDoubleT();
+      }
+
+      auto* public_keys = item.FindListKey("publicKeys");
+      if (!public_keys) {
+        continue;
+      }
+
+      std::string keys_json;
+      base::JSONWriter::Write(*public_keys, &keys_json);
+      promotion->public_keys = keys_json;
+
+      list->push_back(std::move(promotion));
+    }
```

Comment:
Opportunity to break the inner for-loop out into a 'ParsePromotion' method?

Reply:
based on the fact that this PR doesn't modify any of this logic, it's just moves it, I would expend the PR size with refactors